Implement On Demand Downloading of CUDA Dependencies for no CI builds (#9425)
google/iree: CMakeLists.txt

if(IREE_TARGET_BACKEND_CUDA OR IREE_HAL_DRIVER_CUDA)
  find_package(CUDAToolkit)

  if(NOT CUDAToolkit_FOUND AND DEFINED ENV{IREE_CUDA_DEPS_DIR})
    # We define the magic IREE_CUDA_DEPS_DIR env var in our CI docker images if we
    # have a stripped down CUDA toolkit suitable for compiling available. We
    # trigger on this below as a fallback for locating headers and libdevice
    # files.
    set(CUDAToolkit_ROOT "$ENV{IREE_CUDA_DEPS_DIR}")
    message(STATUS "CUDA SDK not found by CMake but using IREE_CUDA_DEPS = ${CUDAToolkit_ROOT}")
  else()

>     # If we haven't found CUDA deps, emulate the CI by downloading them fon demand.

Reviewer: I wouldn't really say we're "emulating the CI" here. We're just making Cuda buildable.

Also, nit: typo

```suggestion
    # If we haven't found CUDA deps, download at least enough to build for Cuda.
```

    # This will define IREE_CUDA_DOWNLOAD_LIBDEVICE_PATH & IREE_CUDA_DOWNLOAD_INCLUDE_PATH
    # vars with the target deps.
    message(STATUS "CUDA SDK not found by CMake but downloading dependencies")

>     add_subdirectory(build_tools/third_party/cuda EXCLUDE_FROM_ALL)

Reviewer: It seems a bit sneaky to do all this work just from including the file. Could we use a function instead?

Author: I can refactor as a function, though it doesn't look like the other `build_tools/third_party` scripts expose functions? They're all included with the `add_subdirectory()`. If I go the function route, should I move this to `build_tools/cmake`? 

To clarify, you want this to be invoked with say a target directory, OS, version, and arch then return the paths to the results? Should target directory, OS, version, and arch be internal details?

Reviewer: My thought was just that "include" is like importing a file, and I normally wouldn't expect that to do something like downloading a file. Some of our other stuff constructs targets (which is also what we do in our own code) which we've modeled as a declarative thing, so the very imperative "go fetch this file" looks weird to me. But I see we have other files doing the same.

I just meant defining a function called "fetch_cuda" or whatever and then calling it explicitly, but I don't think we actually derive any benefit from that, so whatever. Just struck me as weird, but this is CMake, so...
